README.md
# blackjack

> blackjack in javascript

<a href='https://adnjoo.github.io/blackjack/blackjack-v2/index.html'><img src='./preview.png' width=300></a>

Libraries used: [Bootstrap](https://getbootstrap.com/) for styling, [Cards JS](https://github.com/richardschneider/cardsJS) for svg playing cards. [Toastify-js](https://github.com/apvarun/toastify-js) for toasts.

[v1](https://adnjoo.github.io/blackjack/blackjack-v1/index.html)


changelog.md
2021-11-08

add hotkeys for hit, stand, next hand (q, w, e); add toastify-js

rewrite.js
//declare variables: cards, suits
const cards = [
  "A",
  "2",
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "10",
  "J",
  "Q",
  "K",
];
const suits = ["H", "S", "C", "D"];
let deck = {};

// make a deck
function newdeck(rDeck, numberOfDecks = 1) {
  for (let h = 0; h < numberOfDecks; h++) {
    for (let i in cards) {
      for (let j in suits) {
        rDeck[`${cards[i]}${suits[j]}`] = rDeck[`${cards[i]}${suits[j]}`]+1 || 1;
      }
    }
  }
  return rDeck;
}

//shuffle a deck
function shuffle(arr) {
  let newArr = [];
  let count = 0;
  while (newArr.length < arr.length) {
    let sCard = arr[Math.floor(Math.random() * arr.length)];
    for (let i in newArr) {
      if (newArr[i] == sCard) {
        count++;
      }
    }
    if (count < arr.length / 52) {
      newArr.push(sCard);
    }
    count = 0;
  }
  return newArr;
}
deck = newdeck(deck);
// deck = shuffle(deck);
console.log(deck)




scripts/cards.js
/*jslint vars: true, plusplus: true, -W003 */
/*globals window, document, define, exports, module, require */

var cardsScript = document.currentScript;

(function (root, factory) {
    'use strict';

    if (typeof define === 'function' && define.amd) {
        // AMD. Register as an anonymous module.
        define(['jquery'], factory);
    } else if (typeof exports === 'object') {
        // Node. Does not work with strict CommonJS, but
        // only CommonJS-like environments that support module.exports,
        // like Node.
        module.exports = factory(require('jquery'));
    } else {
        // Browser globals (root is window)
        root.cards = factory(root.jQuery);
    }
}(this, function ($) {
    'use strict';

    var module = {
        options: {
            spacing: 0.20,  // How much to show between cards, expressed as percentage of textureWidth
            radius: 400,    // This is the radius of the circle under the fan of cards and thus controls the overall curvature of the fan. Small values means higher curvature
            flow: 'horizontal', // The layout direction (horizontal or vertical)
            fanDirection: "N",
            imagesUrl: "./cards/" // The base URL for the card images, should end with a '/'.
        },

        // Gets the ID of the card, e.g. "KS" for the king of spades.
        cid: function (card) {
            var s = card.attr('src');
            return s.substring(s.length - 6, s.length - 4);
        },

        // Play is called whenever a card in an hand is clicked.  If the hand is active
        // then playCard is called.
        play: function (card) {
            if (card.parents(".active-hand").length > 0) {
                this.playCard(card);
            }
        },

        // Remove a card from the hand.
        remove: function (card) {
            var hand = card.parent();
            card.remove();

            // New layout if card removed from a "fan".
            if (hand.hasClass("fan")) {
                this.fan(hand);
            }
        },

        fan: function (hand, cfg) {
            var options = $.extend({}, this.options),
                cards;

            options = $.extend(options, readOptions(hand, 'fan'));
            if (cfg) {
                options = $.extend(options, cfg);
            }
            hand.data("fan", 'radius: ' + options.radius + '; spacing: ' + options.spacing);
            addCardImages(hand, options.cards);

            cards = hand.find("img.card");
            if (cards.length === 0) {
                return;
            }
            if (options.width) {
                cards.width(options.width);
            }
            fanCards(cards, this, options);
        },

        hand: function ($hand, cfg) {
            var options = $.extend({}, this.options),
                cards,
                width,
                height;
            options = $.extend(options, readOptions($hand, 'hand'));
            if (cfg) {
                options = $.extend(options, cfg);
            }
            $hand.data("hand", 'flow: ' + options.direction + ';');
            $hand.removeClass('hhand fan hhand vhand vhand-compact hhand-compact');
            if (options.flow === 'vertical' && options.spacing >= 1.0) {
                $hand.addClass('vhand');
            } else if (options.flow === 'horizontal' && options.spacing >= 1.0) {
                $hand.addClass('hhand');
            } else if (options.flow === 'vertical') {
                $hand.addClass('vhand-compact');
            } else {
                $hand.addClass('hhand-compact');
            }

            addCardImages($hand, options.cards);

            cards = $hand.find('img.card');
            if (cards.length === 0) {
                return;
            }
            if (options.width) {
                cards.width(options.width);
            }
            width = options.width || cards[0].clientWidth || 70; // hack: for a hidden hand
            height = cards[0].clientHeight || Math.floor(width * 1.4); // hack: for a hidden hand
            if (options.flow === 'vertical' && options.spacing < 1.0) {
                cards.slice(1).css('margin-top', -height * (1.0 - options.spacing));
                cards.slice(1).css('margin-left', 0);
            } else if (options.flow === 'horizontal' && options.spacing < 1.0) {
                cards.slice(1).css('margin-left', -width * (1.0 - options.spacing));
                cards.slice(1).css('margin-top', 0);
            }
        },

        cardSetTop: function (card, top) {
            card.style.top = top + "px";
        },

        cardNames: function (cards) {
            var i,
                name,
                names = [];
            if (typeof cards === 'string') {
                cards = cards.split(' ');
            }
            // Normalise the card names.
            for (i = 0; i < cards.length; ++i) {
                if (cards[i]) {
                    name = cards[i].toString().toUpperCase();
                    names.push(name);
                }
            }

            return names;
        }
    };


    // The default is to remove the card from the hand.
    // module.playCard = module.remove;

    function addCardImages(hand, cards) {
        var i,
            src;
        if (!cards) {
            return;
        }
        cards = module.cardNames(cards);
        hand.empty();
        for (i = 0; i < cards.length; ++i) {
            src = "src='" + module.options.imagesUrl + cards[i] + '.svg' + "'";
            hand.append("<img class='card' " + src + ">");
        }
    }

    // Parse the data-name attribute in HTML.
    function readOptions($elem, name) {
        var v, i, len, s, options, o = {};

        options = $elem.data(name);
        options = (options || '').replace(/\s/g, '').split(';');
        for (i = 0, len = options.length; i < len; i++) {
            s = options[i].split(':');
            v = s[1];
            if (v && v.indexOf(',') >= 0) {
                o[s[0]] = v.split(',');
            } else {
                o[s[0]] = Number(v) || v;
            }
        }
        return o;
    }

    function fanCards(cards, self, options) {
        var n = cards.length;
        if (n === 0) {
            return;
        }

        var width = options.width || cards[0].clientWidth || 90; // hack: for a hidden hand
        var height = cards[0].clientHeight || Math.floor(width * 1.4); // hack: for a hidden hand
        var box = {};
        var coords = calculateCoords(n, options.radius, width, height, options.fanDirection, options.spacing, box);

        var hand = $(cards[0]).parent();
        hand.width(box.width);
        hand.height(box.height);

        var i = 0;
        coords.forEach(function (coord) {
            var card = cards[i++];
            card.style.left = coord.x + "px";
            card.style.top = coord.y + "px";
            card.onmouseover = function () {
                self.cardSetTop(card, coord.y - 10);
            };
            card.onmouseout = function () {
                self.cardSetTop(card, coord.y);
            };
            var rotationAngle = Math.round(coord.angle);
            var prefixes = ["Webkit", "Moz", "O", "ms"];
            prefixes.forEach(function (prefix) {
                card.style[prefix + "Transform"] = "rotate(" + rotationAngle + "deg)" + " translateZ(0)";
            });
        });

    }

    function calculateCoords(numCards, arcRadius, cardWidth, cardHeight, direction, cardSpacing, box) {
        // The separation between the cards, in terms of rotation around the circle's origin
        var anglePerCard = Math.radiansToDegrees(Math.atan(((cardWidth * cardSpacing) / arcRadius)));

        var angleOffset = ({ "N": 270, "S": 90, "E": 0, "W": 180 })[direction];

        var startAngle = angleOffset - 0.5 * anglePerCard * (numCards - 1);

        var coords = [];
        var i;
        var minX = 99999;
        var minY = 99999;
        var maxX = -minX;
        var maxY = -minY;
        for (i = 0; i < numCards; i++) {
            var degrees = startAngle + anglePerCard * i;

            var radians = Math.degreesToRadians(degrees);
            var x = cardWidth / 2 + Math.cos(radians) * arcRadius;
            var y = cardHeight / 2 + Math.sin(radians) * arcRadius;

            minX = Math.min(minX, x);
            minY = Math.min(minY, y);
            maxX = Math.max(maxX, x);
            maxY = Math.max(maxY, y);

            coords.push({ x: x, y: y, angle: degrees + 90 });
        }

        var rotatedDimensions = Math.getRotatedDimensions(coords[0].angle, cardWidth, cardHeight);

        var offsetX = 0;
        var offsetY = 0;

        if (direction === "N") {
            offsetX = (minX * -1);
            offsetX += ((rotatedDimensions[0] - cardWidth) / 2);

            offsetY = (minY * -1);
        } else if (direction === "S") {
            offsetX = (minX * -1);
            offsetX += ((rotatedDimensions[0] - cardWidth) / 2);

            offsetY = ((minY + (maxY - minY)) * -1);
        } else if (direction === "W") {
            offsetY = (minY * -1);
            offsetY += ((rotatedDimensions[1] - cardHeight) / 2);

            offsetX = (minX * -1);
            offsetX += (cardHeight - Math.rotatePointInBox(0, 0, 270, cardWidth, cardHeight)[1]);
        } else if (direction === "E") {
            offsetY = (minY * -1);
            offsetY += ((rotatedDimensions[1] - cardHeight) / 2);

            offsetX = (arcRadius) * -1;
            offsetX -= (cardHeight - Math.rotatePointInBox(0, 0, 270, cardWidth, cardHeight)[1]);
            //offsetX -= ?????;    // HELP! Needs to line up with yellow line!
        }

        coords.forEach(function (coord) {
            coord.x += offsetX;
            coord.x = Math.round(coord.x);

            coord.y += offsetY;
            coord.y = Math.round(coord.y);

            coord.angle = Math.round(coord.angle);
        });

        box.width = coords[numCards - 1].x + cardWidth;
        box.height = coords[numCards - 1].y + cardHeight;

        return coords;
    }

    // If loaded directly from a script, the do the jquery shuffle.
    $(window).on('load', function () {
        // Adjust the cards in a fan, except ones using KO.
        $(".fan:not([data-bind])").each(function () {
            module.fan($(this));
        });

        // Process any data-hand attributes
        $(".hand[data-hand]").each(function () {
            module.hand($(this));
        });

        // Call cards.play, when a card is clicked in an active hand.
        $(".hand").on("click", "img.card", function () {
            module.play($(this));
        });
    });

    // Default imagesUrl to a subfolder of the script source.
    // if (cardsScript && cardsScript.src) {
    //     var path = cardsScript.src.substring(0, cardsScript.src.lastIndexOf('/')) + '/cards/';
    //     module.options.imagesUrl = path;
    // }

    // Just return a value to define the module export.
    // This example returns an object, but the module
    // can return a function as the exported value.
    return module;
}));

// Math Additions
if (!Math.degreesToRadians) {
    Math.degreesToRadians = function (degrees) {
        return degrees * (Math.PI / 180);
    };
}

if (!Math.radiansToDegrees) {
    Math.radiansToDegrees = function (radians) {
        return radians * (180 / Math.PI);
    };
}

if (!Math.getRotatedDimensions) {
    Math.getRotatedDimensions = function (angle_in_degrees, width, height) {
        var angle = angle_in_degrees * Math.PI / 180,
            sin   = Math.sin(angle),
            cos   = Math.cos(angle);
        var x1 = cos * width,
            y1 = sin * width;
        var x2 = -sin * height,
            y2 = cos * height;
        var x3 = cos * width - sin * height,
            y3 = sin * width + cos * height;
        var minX = Math.min(0, x1, x2, x3),
            maxX = Math.max(0, x1, x2, x3),
            minY = Math.min(0, y1, y2, y3),
            maxY = Math.max(0, y1, y2, y3);

        return [ Math.floor((maxX - minX)), Math.floor((maxY - minY)) ];
    };
}

if (!Math.rotatePointInBox) {
    Math.rotatePointInBox = function (x, y, angle, width, height) {
        angle = Math.degreesToRadians(angle);

        var centerX = width / 2.0;
        var centerY = height / 2.0;
        var dx = x - centerX;
        var dy = y - centerY;
        var dist = Math.sqrt(dx * dx + dy * dy);
        var a =  Math.atan2(dy, dx) + angle;
        var dx2 = Math.cos(a) * dist;
        var dy2 = Math.sin(a) * dist;

        return [ dx2 + centerX, dy2 + centerY ];
    };
}

scripts/additional-script.js
class Game {
  constructor(){
    this.bettingstage = false
  }
}
let game = new Game()

class Card {
  constructor(card, suit) {
    this.card = card;
    this.suit = suit;
    this.src = `./cards/${this.card}${this.suit}.svg`;

    //this.value
    if (card == "J" || card == "Q" || card == "K") {
      this.value = 10;
    } else if (card == "A") {
      this.value = card;
    } else {
      this.value = parseInt(card);
    }
  }
}

// make a deck
function newdeck(nDeck, numberOfDecks = 1) {
  for (let h = 0; h < numberOfDecks; h++) {
    for (let i in cards) {
      for (let j in suits) {
        let card = new Card(`${cards[i]}`, `${suits[j]}`);
        nDeck.push(card);
      }
    }
  }
  return nDeck;
}

//shuffle a deck
function shuffle(arr) {
  let newArr = [];
  let count = 0;
  while (newArr.length < arr.length) {
    let sCard = arr[Math.floor(Math.random() * arr.length)];
    for (let i in newArr) {
      if (newArr[i] == sCard) {
        count++;
      }
    }
    if (count < arr.length / 52) {
      newArr.push(sCard);
    }
    count = 0;
  }
  return newArr;
}
deck = newdeck(deck, 2);
deck = shuffle(deck);
console.log(deck, deck.length);


blackjack-v1/script.js
//declare variables: cards, suits
const x = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"];
const suits = ["H", "S", "C", "D"];
let counter = 0;
let deck = {};
let arr = [];
let players = ["player", "dealer"];
let blackjack = 1;
let bettingstage = true;
let winner = null;
let doubled = false;
let myhistory = [];

// make a deck / shuffle the deck
var shuffle = function () {
  for (let i = 0; i < 300; i++) {
    var suit = Math.floor(Math.random() * 4);
    var z = x[Math.floor(Math.random() * 13)].concat(suits[suit]);
    if (!Object.values(deck).includes(z)) {
      deck[i] = z;
    }
  }
  arr = Object.values(deck).sort();
};
shuffle();

// window onload
window.onload = function () {
  //listen to buttons
  document.getElementById("hit").addEventListener("click", function () {
    document.getElementById("double").style.color = "grey";
    if (doubled) {
      return;
    } else if (!bettingstage) {
      hit("player");
    }
  });
  document.getElementById("stand").addEventListener("click", function () {
    if (!bettingstage) {
      stand();
    }
  });
  document.getElementById("nexthand").addEventListener("click", function () {
    if (parseInt(document.getElementById("chips").innerHTML) == 0) {
      return;
    }
    // if 2*betsize is more than chips we grey out the double option
    if (
      2 * parseInt(document.getElementById("betsize").innerHTML) >
      parseInt(document.getElementById("chips").innerHTML)
    ) {
      document.getElementById("double").style.color = "grey";
    }
    if (bettingstage) {
      reset();
      playingstagestyle();
      bettingstage = false;
      hit("player");
      hit("player");
      hit("dealer");
      document.getElementById("hit").classList.remove("hidden");
      document.getElementById("stand").classList.remove("hidden");
      document.getElementById("double").classList.remove("hidden");
      document.getElementById("nexthand").classList.remove("hidden");
      document.getElementById("morechips").classList.remove("hidden");
    }
  });
  document
    .getElementById("increase_bet")
    .addEventListener("click", function () {
      if (bettingstage) {
        increasebet();
      }
    });
  document
    .getElementById("decrease_bet")
    .addEventListener("click", function () {
      if (bettingstage) {
        decreasebet();
      }
      if (parseInt(document.getElementById("betsize").innerHTML) < 100) {
        document.getElementById("betsize").innerHTML = "100";
      }
    });
  document.getElementById("morechips").addEventListener("click", function () {
    if (bettingstage) {
      bettingstagestyle();
    }
    if (parseInt(document.getElementById("betsize").innerHTML) < 100) {
      document.getElementById("betsize").innerHTML = "100";
    }
    document.getElementById("chips").innerHTML = (
      parseInt(document.getElementById("chips").innerHTML) + 100
    ).toString();
  });
  document.getElementById("max_bet").addEventListener("click", function () {
    if (bettingstage) {
      maxbet();
    }
  });

  document.getElementById("double").addEventListener("click", function () {
    // if chips is less than 2 times the betsize, return
    if (
      2 * parseInt(document.getElementById("betsize").innerHTML) >
      parseInt(document.getElementById("chips").innerHTML)
    ) {
      return;
    }
    let divs = document.getElementById("player").children;
    let count = 0;
    for (let i in divs) {
      if (divs[i].localName === "img") {
        count++;
      }
    }
    // if only 2 player cards & chips a.l. 2*betsize
    if (count == 2) {
      document.getElementById("betsize").innerHTML = (
        2 * parseInt(document.getElementById("betsize").innerHTML)
      ).toString();
      doubled = true;
      document.getElementById("hit").style.color = "grey";
      document.getElementById("double").style.color = "grey";
      // if you can double, hit then stand.
      hit("player");
      stand();
    }
  });

  function maxbet() {
    document.getElementById("betsize").innerHTML =
      document.getElementById("chips").innerHTML;
  }

  function increasebet() {
    if (
      bettingstage &&
      parseInt(document.getElementById("betsize").innerHTML) <
        parseInt(document.getElementById("chips").innerHTML)
    ) {
      // if betsize + 100 is more than chips then betsize = chips
      if (
        parseInt(document.getElementById("betsize").innerHTML) + 100 >
        parseInt(document.getElementById("chips").innerHTML)
      ) {
        document.getElementById("betsize").innerHTML =
          document.getElementById("chips").innerHTML;
        return;
      }
      document.getElementById("betsize").innerHTML = (
        parseInt(document.getElementById("betsize").innerHTML) + 100
      ).toString();
    }
  }

  function decreasebet() {
    if (
      bettingstage &&
      parseInt(document.getElementById("betsize").innerHTML) > 100
    ) {
      document.getElementById("betsize").innerHTML = (
        parseInt(document.getElementById("betsize").innerHTML) - 100
      ).toString();
    }
  }
};
// hits player or dealer
function hit(who) {
  if (bettingstage) {
    return;
  }
  counter = Math.abs(arr.length - 52);
  var h = arr[Math.floor(Math.random() * arr.length)];
  myhistory.push(`${who} ${h}`);
  arr = arr.filter((e) => e !== h);
  //create object card to append to DOM
  var card = document.createElement("img");
  card.classList.add("playingcard");
  card.id = "card" + counter.toString();
  card.id += ` ${h}`;
  card.src = `cards/${h}.svg`;
  document.getElementById(who).appendChild(card);
  var y = getscoreofcards(who, null);
  // if score is greater than 21 then bust the who
  if (y > 21) {
    document.getElementById(who + "bust").innerHTML = "bust";
    if (who == "player") {
      winner = "dealer";
      endofgame(true);
    } else {
      winner = "player";
      endofgame(true);
    }
    bettingstage = true;
  }
  document.getElementById(`${who}score`).innerHTML = y.toString();
}

// pass cards in (of who) and get score out
function getscoreofcards(who, final) {
  var cards = returnhand(who);
  var total = 0;
  var acecounter = 0;
  for (let i in cards) {
    //convert J, Q, K to 10
    if (cards[i] == "J" || cards[i] == "Q" || cards[i] == "K") {
      cards[i] = 10;
    }
    if (cards[i] != "A") {
      cards[i] = parseInt(cards[i]);
    }
  }
  // if it doesn't contain A -> calc score
  if (!cards.includes("A")) {
    for (let i in cards) {
      total += cards[i];
    }
    return total;
  }
  // if it contains 1 or more Ace.
  //check if who is player, and one ace, and one ten card
  else {
    for (let i in cards) {
      if (cards[i] != "A") {
        total += cards[i];
      }
      if (cards[i] == "A") {
        acecounter++;
      }
    }
    if (acecounter == 1 && cards.length == 1) {
      return 11;
    }
    if (
      acecounter == 1 &&
      total == 10 &&
      cards.length == 2 &&
      who == "player"
    ) {
      blackjack = 1.5;
    }
    // return 11 + total, or all 1s and total
    var a = acecounter + total;
    var b = 10 + total + acecounter;
    // if the higher total is 21 we want to return 21.
    if (b == 21) {
      return 21;
    } else if (b > 21) {
      return a;
    }
    // if it's the dealer's final hand we want to return the highest number available to the dealer.
    if (who == "dealer") {
      return b;
    }
    if (final) {
      return b;
    }
    // if there's more than 1 ace we return the soft Aces
    // + total, as well as the one hard Ace (11) and soft Ace(s), + total.
    return "total is " + a.toString() + " or " + b.toString();
  }
}

//function to return hand of player || dealer
function returnhand(who) {
  var x = document.getElementById(who).children;
  var cards = [];
  for (let i in x) {
    //get the cards first
    if (x[i].localName == "img") {
      cards.push(x[i].id);
    }
  }
  for (let i in cards) {
    // if its 10 remove last letter/ suit
    if (cards[i][cards[i].length - 3] == 1) {
      cards[i] = cards[i].substring(cards[i].length - 3, cards[i].length - 1);
    } else {
      //remove last letter/ suit
      cards[i] = cards[i].substring(cards[i].length - 2, cards[i].length - 1);
    }
  }
  return cards;
}

//dealer's gameplay
function stand() {
  check17();
  setTimeout(function () {
    endofgame(null);
  }, 1500);
}

//dealer hits to 17 using recursive call
function check17() {
  if (getscoreofcards("dealer", null) > 16) {
    return;
  } else {
    hit("dealer");
    setTimeout(function () {
      check17();
    }, 400);
  }
}

//resets game stage style to betting
function bettingstagestyle() {
  document.getElementById("hit").style.color = "grey";
  document.getElementById("stand").style.color = "grey";
  document.getElementById("nexthand").style.color = "black";
  document.getElementById("nexthand").style.fontWeight = "bold";
  document.getElementById("increase_bet").style.color = "black";
  document.getElementById("decrease_bet").style.color = "black";
  document.getElementById("max_bet").style.color = "black";
  document.getElementById("double").style.color = "grey";
}

//sets game stage style to playing
function playingstagestyle() {
  document.getElementById("hit").style.color = "white";
  document.getElementById("stand").style.color = "white";
  document.getElementById("nexthand").style.color = "grey";
  document.getElementById("increase_bet").style.color = "red";
  document.getElementById("decrease_bet").style.color = "red";
  document.getElementById("max_bet").style.color = "red";
  document.getElementById("double").style.color = "white";
  // betsize*2 > chips, double style is grey
  if (
    2 * parseInt(document.getElementById("betsize").innerHTML) >
    parseInt(document.getElementById("chips").innerHTML)
  ) {
    document.getElementById("double").style.color = "grey";
  }
}

//delete all cards and reset variables
function reset() {
  //delete all cards in play (divs)
  var x = document.getElementById("player").children;
  for (let i = 0; i < x.length; i++) {
    if (x[i].localName == "img") {
      var yy = document.getElementById(x[i].id);
      yy.remove();
      i--;
    }
  }
  var a = document.getElementById("dealer").children;
  for (let i = 0; i < a.length; i++) {
    if (a[i].localName == "img") {
      var y = document.getElementById(a[i].id);
      y.remove();
      i--;
    }
  }
  //reset style and elements
  document.getElementById("winner").innerHTML = "";
  document.getElementById("playerbust").innerHTML = "";
  document.getElementById("dealerbust").innerHTML = "";
  playingstagestyle();

  // reset values
  blackjack = 1;
  doubled = false;
}

//check winner
function endofgame(bust) {
  //change style
  bettingstagestyle();
  //make a new deck
  if (arr.length < 10) {
    shuffle();
  }
  //if player busts, and dealer wins
  if (bust && winner == "dealer") {
    losemoney();
  }
  // if dealer busts we check for blackjack
  else if (bust && winner == "player") {
    checkblackjack();
  }
  // if dealer wins and is not bust
  else if (
    getscoreofcards("dealer", null) > getscoreofcards("player", true) &&
    getscoreofcards("dealer", null) < 22
  ) {
    losemoney();
  }
  // if dealer hits to 17 or over but is less than player
  else if (getscoreofcards("dealer", null) < getscoreofcards("player", true)) {
    checkblackjack();
  }
  // if they are tied.
  else if (getscoreofcards("dealer", null) == getscoreofcards("player", true)) {
    document.getElementById("winner").innerHTML = "push - tie";
  }
  // resetting betsize amount if it's lower than the chips the player has
  if (
    parseInt(document.getElementById("betsize").innerHTML) >
    parseInt(document.getElementById("chips").innerHTML)
  ) {
    document.getElementById("betsize").innerHTML =
      document.getElementById("chips").innerHTML;
  }
  bettingstage = true;
}

//dealer wins; player loses
function losemoney() {
  document.getElementById("winner").innerHTML = "dealer wins!";
  document.getElementById("chips").innerHTML = (
    parseInt(document.getElementById("chips").innerHTML) -
    parseInt(document.getElementById("betsize").innerHTML)
  ).toString();
  // betsize is lower of betsize e.g. 300, or default bet 100
  if (parseInt(document.getElementById("chips").innerHTML) < 200) {
    document.getElementById("betsize").innerHTML = "100";
  }
  //set styling if chips are worth 0
  if (parseInt(document.getElementById("chips").innerHTML) <= 0) {
    document.getElementById("nexthand").style.color = "grey";
  }
}

//player wins so check blackjack
function checkblackjack() {
  document.getElementById("winner").innerHTML = "player wins!";
  if (blackjack == 1.5) {
    document.getElementById("winner").innerHTML = "blackjack!";
    document.getElementById("chips").innerHTML = (
      parseInt(document.getElementById("chips").innerHTML) +
      blackjack * parseInt(document.getElementById("betsize").innerHTML)
    ).toString();
  }
  if (blackjack == 1) {
    document.getElementById("chips").innerHTML = (
      parseInt(document.getElementById("chips").innerHTML) +
      parseInt(document.getElementById("betsize").innerHTML)
    ).toString();
  }
  bettingstage = true;
}
